lib/Dialect/TritonGPU/Transforms/WarpSpecialization/LoadMMASpecialization.cpp

```diff
@@ -137,90 +138,364 @@ static PartitionScheme assignPartitions(scf::ForOp loop) {
+      uses.push_back(&use);
+  }
 }
 
-static WarpSchedule getInitialSchedule(const PartitionScheme &scheme) {
+// Given a partitioning scheme, determine an initial schedule by performing a
+// first-order partition assignment to the operations in the scheme and its
+// users and/or dependencies. This sets up the initial partitioning of the ops.
+static WarpSchedule getInitialSchedule(const PartitionScheme &scheme,
+                                       scf::ForOp loop) {
   WarpSchedule schedule;
 
+  // Start by creating the default partition, a partition for for all loads, and
+  // a partition for all MMAs.
+  Partition *defaultPartition = schedule.addPartition(0);
+  Partition *mmaPartition = schedule.addPartition(1);
   Partition *loadPartition = schedule.addPartition(0);
+
   for (const PipelinedLoad &load : scheme.loads) {
-    loadPartition->insert(load.loadOp);
+    schedule.trySchedule(loadPartition, load.loadOp);
     for (Operation *allocOp : load.allocOps)
-      loadPartition->insert(allocOp);
+      schedule.trySchedule(loadPartition, allocOp);
   }
 
-  Partition *mmaPartition = schedule.addPartition(1);
   for (const PipelinedMMA &mma : scheme.mmas) {
-    mmaPartition->insert(mma.mmaOp);
+    schedule.trySchedule(mmaPartition, mma.mmaOp);
     if (mma.storeOp)
-      mmaPartition->insert(mma.storeOp);
+      schedule.trySchedule(mmaPartition, mma.storeOp);
     for (Operation *viewOp : mma.operandViews)
-      mmaPartition->insert(viewOp);
+      schedule.trySchedule(mmaPartition, viewOp);
   }
 
-  if (!scheme.userOps.empty()) {
-    Partition *userPartition = schedule.addPartition(2);
-    for (Operation *userOp : scheme.userOps)
-      userPartition->insert(userOp);
-    // Place the epilogue partition in the default warpgroup. The MMA and load
-    // partitions shouldn't have tensor computations in them, which means they
-    // will get assigned just 1 warp each.
-    schedule.reorderPartitions({2, 1, 0});
+  // Propagate defs of exp.
+  for (math::Exp2Op exp : scheme.exps) {
+    schedule.trySchedule(defaultPartition, exp);
+    scheduleDependencies(loop, schedule, defaultPartition, exp);
+  }
+
+  // Propagate users of loads and MMAs.
+  for (const PipelinedLoad &load : scheme.loads) {
+    scheduleUsers(loop, schedule, defaultPartition, load.loadOp);
+    for (Operation *allocOp : load.allocOps)
+      scheduleUsers(loop, schedule, defaultPartition, allocOp);
+  }
+
+  SmallVector<Partition *> userPartitions{defaultPartition};
+  while (userPartitions.size() < scheme.mmas.size()) {
+    userPartitions.push_back(schedule.addPartition(userPartitions.size()));
+  }
+  for (auto [mma, userPartition] : llvm::zip(scheme.mmas, userPartitions)) {
+    scheduleUsers(loop, schedule, userPartition, mma.mmaOp);
+  }
+  for (const PipelinedMMA &mma : scheme.mmas) {
+    scheduleDependencies(loop, schedule, defaultPartition, mma.mmaOp);
   }
 
   schedule.updatePartitions();
   return schedule;
 }
 
+namespace {
```

Comment:
I'd love a little more description on the OpClusters and the propogate partitions function. This seems to be the crux of the PR, and I'm not sure I understand it yet.